File: app/src/main/java/agersant/polaris/api/local/LocalAPI.java
package agersant.polaris.api.local;

import android.graphics.Bitmap;

import com.google.android.exoplayer2.source.MediaSource;

import java.io.IOException;
import java.util.ArrayList;

import agersant.polaris.CollectionItem;
import agersant.polaris.api.IPolarisAPI;
import agersant.polaris.api.ItemsCallback;
import agersant.polaris.api.PlaylistsCallback;
import agersant.polaris.api.ThumbnailSize;


public class LocalAPI implements IPolarisAPI {

    private OfflineCache offlineCache;

    public LocalAPI() {
    }

    public void initialize(OfflineCache offlineCache) {
        this.offlineCache = offlineCache;
    }

    public boolean hasAudio(CollectionItem item) {
        String path = item.getPath();
        return offlineCache.hasAudio(path);
    }

    @Override
    public MediaSource getAudio(CollectionItem item) throws IOException {
        String path = item.getPath();
        return offlineCache.getAudio(path);
    }

    public boolean hasImage(CollectionItem item, ThumbnailSize size) {
        String path = item.getArtwork();
        return offlineCache.hasImage(path, size);
    }

    public Bitmap getImage(CollectionItem item, ThumbnailSize size) {
        String artworkPath = item.getArtwork();
        try {
            return offlineCache.getImage(artworkPath, size);
        } catch (IOException e) {
            System.out.println("Error while retrieving image from local cache: " + artworkPath);
        }
        return null;
    }

    @Override
    public void browse(String path, ItemsCallback handlers) {
        ArrayList<CollectionItem> items = offlineCache.browse(path);
        if (items == null) {
            handlers.onError();
        } else {
            handlers.onSuccess(items);
        }
    }

    @Override
    public void flatten(String path, ItemsCallback handlers) {
        ArrayList<CollectionItem> items = offlineCache.flatten(path);
        if (items == null) {
            handlers.onError();
        } else {
            handlers.onSuccess(items);
        }
    }

    @Override
    public void search(String query, ItemsCallback handlers) {
        ArrayList<CollectionItem> items = offlineCache.search(query);
        if (items == null) {
            handlers.onError();
        } else {
            handlers.onSuccess(items);
        }
    }

    @Override
    public void getPlaylists(PlaylistsCallback handlers) {
        throw new UnsupportedOperationException("Not yet implemented"); // TODO implement offline playlists
    }

    @Override
    public void getPlaylist(String name, ItemsCallback handlers) {
        throw new UnsupportedOperationException("Not yet implemented"); // TODO implement offline playlists
    }
}

File: app/src/main/java/agersant/polaris/api/local/LocalAPI.kt
package agersant.polaris.api.local

import agersant.polaris.CollectionItem
import agersant.polaris.Song
import agersant.polaris.api.IPolarisAPI
import agersant.polaris.api.ThumbnailSize
import android.graphics.Bitmap
import com.google.android.exoplayer2.source.MediaSource
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.withContext

class LocalAPI : IPolarisAPI {
    private lateinit var offlineCache: OfflineCache

    fun initialize(offlineCache: OfflineCache) {
        this.offlineCache = offlineCache
    }

    fun hasAudio(item: Song): Boolean {
        val path = item.path
        return offlineCache.hasAudio(path)
    }

    override suspend fun getAudio(item: Song): MediaSource? = withContext(IO) {
        offlineCache.getAudio(item.path)
    }

    fun hasImage(item: CollectionItem, size: ThumbnailSize): Boolean {
        val path = item.artwork ?: return false
        return offlineCache.hasImage(path, size)
    }

    fun getImage(item: CollectionItem, size: ThumbnailSize): Bitmap? {
        val path = item.artwork ?: return null
        return offlineCache.getImage(path, size)
    }

    override suspend fun browse(path: String): List<CollectionItem>? = withContext(IO) {
        offlineCache.browse(path)
    }

    override suspend fun flatten(path: String): List<Song>? = withContext(IO) {
        offlineCache.flatten(path)
    }
}
